src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,164 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <variant>
+#include <optional>
+
+
+#include "antlr4-runtime.h"
+#include "absl/status/statusor.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/ast_construction.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h" 
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+Principal Parser::ConstructPrincipal (AuthLogicParser::PrincipalContext* context) {
+    std::string name = context->ID()->getText();
+    return Principal(name);
+}
+
+datalog::Predicate Parser::ConstructPredicate (AuthLogicParser::PredicateContext* context ) {
+    std::string name = context->ID()->getText();
+    std::vector<std::string> args;
+    datalog::Sign sign;
+    
+    for (auto& statement : context->pred_arg()) {
+        args.push_back(std::move(statement->ID()->getText()));
+    }
+
+    sign = context->NEG() == nullptr ? datalog::Sign::kPositive : datalog::Sign::kNegated;
```

Comment:
Is it the case that `NEG()` will only return `nullptr` or something else? Can it be explicitly specified as positive?

Reply:
Right, NEG can only be nullptr or something else. This is the part of the grammar for negating a predicate, so the negation symbol is either in the code that was parsed or it wasn't.

Reply:
Maybe it's safer to call `getText()` when `NEG` is not a `nullptr` and set `sign = datalog::Sign::kNegated` only when the text is the right value (`!`) and throw an exception when NEG is some other non-null value. I think this is also fine as it is, though.

Reply:
I am going to keep this as it is. antlr parser would throw a RecognitinException error for illegal input text.